### PR TITLE
GH-84976: Create Lib/_pydatetime to optimize import datetime when C implementation is available

### PR DESCRIPTION
**Pull request by pganssle**
From the discussion on [stack overflow](https://stackoverflow.com/questions/75825183/is-there-a-way-to-move-a-file-and-then-change-it-in-such-a-way-that-that-git-can), #84976 and #99090, no one can find a way to do this that allows clean cherry-picks, but also no one can explain why it's *not* possible, but at this point it's not worth delaying any more. We rarely need to backport things into `datetime.py`, and in those rare cases we can do it manually. If it suddenly becomes very common, we can maybe special-case it in Miss Islington.

* Issue: gh-84976

### CODE REVIEW

**Comment by pganssle** [comment]
CC @ezio-melotti

**Comment by pganssle** [comment]
Even though it doesn't seem to help us cherry-pick, I would feel a bit more comfortable if I could merge this into `main` as-is rather than doing a squash merge.

@Yhg1s As RM for 3.12, I *think* you are the right person to ask — is it possible to merge this into `main` without squash merge? I think it would be 3 or 4 commits instead of 1.

**Comment by Yhg1s** [comment]
FYI, I've merged the two commits with rebase (not squashing), so they are two separate commits in the main branch, which hopefully makes future backports easier.

File: Lib/datetime.py [hunk 1]
```diff
@@ -1,2663 +1,9 @@
...
-    # docstring does not get overwritten. In the future, it may be
-    # appropriate to maintain a single module level docstring and
-    # remove the following line.
     from _datetime import __doc__
+except ImportError:
+    from _pydatetime import *
+    from _pydatetime import __doc__
+
+__all__ = ("date", "datetime", "time", "timedelta", "timezone", "tzinfo",
+           "MINYEAR", "MAXYEAR")
```
Comment (ngnpope) [inline Lib/datetime.py:9]: @pganssle It looks like `UTC` got dropped accidentally as the original version of this patch predated its addition?
```suggestion
           "MINYEAR", "MAXYEAR", "UTC")
```

Comment (hauntsaninja) [reply Lib/datetime.py:9]: Thanks, I noticed this as well. Fixed in https://github.com/python/cpython/pull/104920

**Pull request merged**